test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java

```diff
@@ -0,0 +1,423 @@
+        assertEquals(eventData.get(), new TreeSet<>(eventsReadFromPravega).size()); //check unique events.
+        log.debug("{} with 2 controller, 2 SSS instances succeeds", "multiReaderWriterTest");
+
+        //scale down 1 SSS instance
+        segmentStoreInstance.scaleService(2, true);
+        Thread.sleep(60000);
+        log.info("Scaling down  1 SSS instance");
+
+        eventsReadFromPravega = new ConcurrentLinkedQueue<>();
+        stopReadFlag.set(false);
+        eventData = new AtomicLong(); //data used by each of the writers.
+        eventReadCount = new AtomicLong(); // used by readers to maintain a count of events.
+
+        // start writing asynchronously
+        CompletableFuture<Void> writerFuture1 = CompletableFuture.runAsync(() -> {
+            writerList.forEach(writer -> startWriting(eventData, writer));
+        });
+        //start reading asynchronously
+        CompletableFuture<Void> readerFuture1 = CompletableFuture.runAsync(() -> {
+            readerList.forEach(reader -> startReading(eventsReadFromPravega, eventData, eventReadCount, stopReadFlag, reader));
+        });
+        //wait for writers completion
+        writerFuture1.get();
+        //wait till #.reads = #.writes
+        while (NUM_EVENTS != eventsReadFromPravega.size()) {
+            Thread.sleep(5);
+        }
+        //stop reading when no. of reads= no. of writes
+        stopReadFlag.set(true);
+        //wait for readers completion
+        readerFuture1.get();
+        log.info("All writers have stopped. Setting Stop_Read_Flag. Event Written Count:{}, Event Read " +
+                "Count: {}", eventData.get(), eventsReadFromPravega.size());
+        assertEquals(eventData.get(), eventsReadFromPravega.size());
+        assertEquals(eventData.get(), new TreeSet<>(eventsReadFromPravega).size()); //check unique events.
+        log.debug("{} with SSS failover succeeds", "multiReaderWriterTest");
+
+        //scale down 1 controller instance
+        controllerInstance.scaleService(2, true);
+        List<URI> conURIs = controllerInstance.getServiceDetails();
+        // Fetch all the RPC endpoints and construct the client URIs.
+        final List<String> uris = conURIs.stream().filter(uri -> uri.getPort() == 9092).map(URI::getAuthority)
+                .collect(Collectors.toList());
+        controllerURIDirect = URI.create("tcp://" + String.join(",", uris));
+        log.info("Controller Service direct URI: {}", controllerURIDirect);
+        Thread.sleep(60000);
+        log.info("Test with controller failover");
+
+        eventsReadFromPravega = new ConcurrentLinkedQueue<>();
+        stopReadFlag.set(false);
+        eventData = new AtomicLong(); //data used by each of the writers.
+        eventReadCount = new AtomicLong(); // used by readers to maintain a count of events.
+
+        // start writing asynchronously
+        CompletableFuture<Void> writerFuture2 = CompletableFuture.runAsync(() -> {
```

Comment:
The readers and writers test code is repeated for every scenario, this code can be reused by moving to a common function.